Show exception details in text visualizer dialog

drewnoakes opened the pull request:
## Description

Fixes #6130

Exception details are shown behind a button in two places:

1. In the "Message" column of the structured logs table, when an exception exists.
2. In the "Health reports" grid in resource details, when a health report includes exception details.

Previously, the stack trace would be shown in a popup when the mouse hovered over the icon, sort of like a tooltip.

This commit changes the behaviour so that exception details are shown in the existing text visualizer dialog whenever the icon is clicked.

The icon becomes a `FluentButton` for consistency with the menu button to the right of the cell.

Also some optimisations in `GridValue`:

- Don't use `FluentHighlighter` when the text to highlight is empty, which will be the common case.
- Remove `MaxDisplayLength`. We never set this property, and it used space on every `GridValue` instance, and we create a lot of these objects.
- Remove some redundant DOM elements
- Support adding arbitrary buttons to the right-hand side of the area (used for the "exception details" button)

---

Here's how it looks for health checks:

![image](https://github.com/user-attachments/assets/40c1a534-fca5-40f4-837e-4f9d33c77ce0)

Here's how it looks on mouse-over, along with a tooltip. The button appears too close to the text here, but that's consistent with other inline lightweight buttons. It only shows on mouse over. If we add more margin, it looks bad when the mouse isn't hovering over it.

![image](https://github.com/user-attachments/assets/256c61a4-3b11-46f4-867c-36a726ed3cb9)

Clicking the button opens the text visualizer we use throughout the dashboard:

![image](https://github.com/user-attachments/assets/4b83d32f-4555-43f8-a579-b0436cb09551)

Here's how it looks in structured logs:

![image](https://github.com/user-attachments/assets/042764c2-7e4a-422f-a38c-45d1706df2dc)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6230)

davidfowl commented:
Yes please!

davidfowl commented:
/azp run

JamesNK commented on src/Aspire.Dashboard/Components/Controls/ExceptionDetails.razor:
```diff
@@ -1,52 +1,35 @@
-            </div>
-        </div>
-    </FluentTooltip>
+    <FluentButton Appearance="Appearance.Lightweight"
```
It doesn't look good on mouse over with the button hard up against the text:

![image](https://github.com/user-attachments/assets/179011ab-1d8e-4561-87fb-0b351d7b7b10)

Reduce the width of the button (maybe have 4px?) and add a right margin to provide some space between the button and text.

Example from the inline error button on resources page:

https://github.com/dotnet/aspire/blob/fe53c9874e5db7e64be5221b5f41c0ea1d9203b0/src/Aspire.Dashboard/Components/Pages/Resources.razor.css#L6-L11

drewnoakes replied:
I called this out in:

> Here's how it looks on mouse-over, along with a tooltip. The button appears too close to the text here, but that's consistent with other inline lightweight buttons. It only shows on mouse over. If we add more margin, it looks bad when the mouse isn't hovering over it.

Making this button smaller means it no longer matches the equivalent button on the right. We should be consistent.

JamesNK replied:
Do you mean the `...` button that appear on hover? I wouldn't call it inline. It appears over the top of content on hover. The existing prior art of a clickable icon that is inline in text is the unread logs errors button. It doesn't have padding, just a margin to give it space from text (CSS above).

The problem with having the large padding is content isn't consistently aligned with the start of the column. It doesn't look good. And it breaks consistency with other columns that start with an icon (e.g. the level column also in the structured logs table).

Aspire today:

![image](https://github.com/user-attachments/assets/ac3ea25c-3291-4552-a045-3c3940d82494)

Current PR:

![image](https://github.com/user-attachments/assets/193b5721-8fcb-4d6c-8941-c30df7ad68f2)

Appearance wise it should stay the same, with the change in behavior from hover to click.

cc @davidfowl @leslierichardson95 for input.

davidfowl replied:
Honestly, I would prefer if we removed that little box and moved it to where the actions are on the left. Similar to how start/stop look on the resources view.

JamesNK replied:
Hmm, that's an interesting idea. One problem I see with that is if there is an error then the `...` actions button would have to always be shifted right to make room for the "View exception details" button. But a lot of the time there won't be any logs exception details visible in the UI, so the `...` button will look like it is in the wrong place.

Here is an alternative idea:

What about changing the button in this PR to just a regular icon. No hovering, no clicking. Then in the actions menu, there is a "Exception details" option. There is already an menu option for viewing the message in the text visualizer. Would be simple to do the same thing for the exception details.

Imagine an extra option for "Exception details" with its matching icon in the picture below:
![image](https://github.com/user-attachments/assets/8d596991-e463-4e68-98e1-1e4c6f748483)

davidfowl replied:
Prefer if it was aligned. A button without a tooltip feels bad though. It's hard to tell what that icon even is until you hover over it. @DamianEdwards was lamenting that the experience of viewing errors in the structured logs felt difficult compared to looking at console logs.

I think the core fix here is to add the menu item to view the exception details in the grid view *and* in the details view. If we can do the same with the icon then great, but its less discoverable IMO.

DamianEdwards replied:
How about changing the status icon instead of resources with unread errors, rather than trying to fit an extra icon in somewhere else?

drewnoakes replied:
I moved the button to the right hand side of the "cell" so that it sits alongside the masking button and the ... dropdown menu button. There's a generic facility for us to add other buttons there in future.

drewnoakes replied:
See updated screenshots in description.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/ExceptionDetails.razor:
```diff
@@ -1,52 +1,35 @@
-            </div>
-        </div>
-    </FluentTooltip>
+    <FluentButton Appearance="Appearance.Lightweight"
```
Clicking on the button opens and closes the log details because the click event propagates to the row click handle. Stop that with stopPropagation. Example:

https://github.com/dotnet/aspire/blob/fe53c9874e5db7e64be5221b5f41c0ea1d9203b0/src/Aspire.Dashboard/Components/ResourcesGridColumns/UnreadLogErrorsBadge.razor#L12

If the attribute doesn't work on the button, then add a span/div around the button with the attribute.

drewnoakes replied:
Moved to a container for all buttons in `GridValue`, so it applies consistently.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/GridValue.razor.css:
```diff
@@ -9,19 +9,23 @@
     white-space: nowrap;
 }
 
+::deep .grid-value > .exception-details-button {
```
IMO add `exception-details-button` styles to global app.css.

The component isn't specific to grid value - it could be placed anywhere - it just happens to only be used in grid value at the moment.

drewnoakes replied:
Further changes made this rule unnecessary.

davidfowl commented:
I would love this in 9

drewnoakes commented:
@JamesNK @davidfowl @DamianEdwards I moved the icon to the right side, so it sits with the masking button and the "..." dropdown button. See updated screenshots in PR description.

JamesNK reviewed (approved):
I like it.

One extra suggestion is to add an action to the menu here:
![image](https://github.com/user-attachments/assets/08306411-668a-4014-b81c-cf475ff09bca)

If there is an exception message then add a menu item with `<icon> Exception details` and launches the same visualizer. The `Log message` menu item already does the same thing.

If it is difficult then create a follow up issue. Stick with this for now and add the menu item in 9.1.

drewnoakes commented:
Adding a menu item makes sense, though it might reasonably be expected in the `GridValue`'s own drop-down menu. That'd make it work in the Health Checks area too. Any thoughts?

I was tempted to pull out an abstraction for these buttons and menu items, to take a more data-driven approach. I would rather do that in 9.1, to keep the scale of the change down.

JamesNK commented:
I found a regression:
* Previously the content in the grid cell would use the full width. Content would shrink when needed to make room for the `...` button
* Now the content is always strunk to make room for the `...` button.

Before:
![log-button-main](https://github.com/user-attachments/assets/d11aa89a-bf83-4f9f-9faa-d99d6478037e)

After:
![log-button-branch](https://github.com/user-attachments/assets/2ed4bc4c-90b5-4a5f-8f35-2c5f2b4f4a3d)

JamesNK commented:
> Adding a menu item makes sense, though it might reasonably be expected in the `GridValue`'s own drop-down menu. That'd make it work in the Health Checks area too. Any thoughts?

We'd need to add a way to add items to the grid value menu. It's doable, and I like the consistency, but not as high value. Structured logs will get a lot of use.

Could do it properly in 9.1.

drewnoakes commented:
> Previously the content in the grid cell would use the full width. Content would shrink when needed to make room for the ... button. Now the content is always strunk to make room for the ... button.

Fixed in 340124ae04b989794cb14bb7d01f4a031d7f9170.

drewnoakes commented:
@JamesNK what do you think of this: 

![image](https://github.com/user-attachments/assets/283e7495-4c93-47cc-87d6-108b0c14ee23)

The filled icon looks a bit strange, but there's no regular version of that icon as far as I can tell.

JamesNK commented:
* I prefer the text as `Exception details`. That's consistent with `Log message`. I don't know why I like this system, but not having `View details` or consistently having `View xxx` for all items felt worse.
* Move `Log details` above `Exception details`. That way there is consistent order of items. Always `View details` and `LogMessage` then an optional `Exception details`

drewnoakes committed:
Show exception details in text visualizer dialog

Exception details are shown behind a button in two places:

1. In the "Message" column of the structured logs table, when an exception exists.
2. In the "Health reports" grid in resource details, when a health report includes exception details.

Previously, the stack trace would be shown in a popup when the mouse hovered over the icon, sort of like a tooltip.

This commit changes the behaviour so that exception details are shown in the existing text visualizer dialog whenever the icon is clicked.

The icon becomes a `FluentButton` for consistency with the menu button to the right of the cell.

Also some optimisations in `GridValue`:

- Don't use `FluentHighlighter` when the text to highlight is empty, which will be the common case. This improves UI performance.
- Remove `MaxDisplayLength`. We never set this property, and it used space on every `GridValue` instance, and we create a lot of these objects.
- Remove some redundant DOM elements.
- Support adding arbitrary buttons to the right-hand side of the area (used for the "exception details" button).

drewnoakes commented:
I'll do the menu item in a follow-up PR for 9.1.

drewnoakes commented:
/backport to release/9.0
